core/src/main/scala/kafka/server/ControllerApis.scala

```diff
@@ -220,12 +221,64 @@ class ControllerApis(val requestChannel: RequestChannel,
   def handleBrokerRegistration(request: RequestChannel.Request): Unit = {
     val registrationRequest = request.body[BrokerRegistrationRequest]
     apisUtils.authorizeClusterOperation(request, CLUSTER_ACTION)
-    controller.registerBroker(registrationRequest.data())
+
+    controller.registerBroker(registrationRequest.data).handle[Unit]((reply, ex) => {
+      var err = Errors.NONE
+      if (ex != null) {
+        error(s"Failed to register broker: ${ex.getMessage}")
+        err = Errors.forException(ex)
+      }
+
+      def createResponseCallback(requestThrottleMs: Int, reply: RegistrationReply, error: Errors): BrokerRegistrationResponse = {
+        val brokerRegistrationResponseData = new BrokerRegistrationResponseData()
+
+        // TODO: get a better estimate of the current active controller from Raft or somewhere.
+        if (controller.curClaimEpoch() > 0) {
+          brokerRegistrationResponseData.setActiveControllerId(config.controllerId)
+        } else {
+          brokerRegistrationResponseData.setActiveControllerId(MetadataResponse.NO_CONTROLLER_ID)
+        }
+        brokerRegistrationResponseData.setBrokerEpoch(reply.epoch)
+        brokerRegistrationResponseData.setLeaseDurationMs(reply.epoch)
```

Comment:
lease duration shouldn't be set as epoch

Reply:
Good catch!